kernel/cClosure.ml

```diff
@@ -260,7 +266,7 @@ let zupdate info m s =
   let share = info.i_cache.i_share in
   if share && is_red m.mark then
     let s' = compact_stack m s in
-    let _ = m.term <- FLOCKED in
+    let _ = m.term <- FLOCKED; m.as_constr <- None in
```

Comment:
@Janno I guess this is what you mean by "discard the term when it's changed"

Reply:
My understanding of the mutable parts of CClosure is still very bad. This certainly seems like one important point where any cached `constr` value would need to be discarded. I was under the impression that there were places were `fconstr`s were constructed from existing `fconstr`s but that does not seem to be the case. So maybe this is indeed the only place that needs to reset to `None`.